src/Data/BoolExpr.hs

```diff
@@ -0,0 +1,337 @@
+A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE NoDeriveAnyClass #-}
+
+--------------------------------------------------------------------
+
+--------------------------------------------------------------------
+
+-- |
+-- Module    : Data.BoolExpr
+-- Copyright : (c) Nicolas Pouillard 2008,2009
+-- License   : BSD3
+--
+-- Maintainer: Nicolas Pouillard <nicolas.pouillard@gmail.com>
+-- Stability : provisional
+-- Portability:
+--
+-- Boolean expressions and various representations.
+module Data.BoolExpr (
+  -- * A boolean class
+  Boolean (..),
+
+  -- * Boolean trees
+  BoolExpr (..),
+  reduceBoolExpr,
+  evalBoolExpr,
+
+  -- * Boolean evaluation semantic
+  Eval (..),
+  runEvalId,
+
+  -- * Signed constants
+  Signed (..),
+  negateSigned,
+  evalSigned,
+  constants,
+  negateConstant,
+
+  -- * Conjunctive Normal Form
+  CNF (..),
+  Conj (..),
+  boolTreeToCNF,
+
+  -- * Disjunctive Normal Form
+  Disj (..),
+  DNF (..),
+  boolTreeToDNF,
+
+  -- * Other transformations
+  dualize,
+  fromBoolExpr,
+  pushNotInwards,
+)
+where
+
+import Control.Monad (ap)
+import Data.Aeson
+import Data.Char (toLower)
+import Data.Traversable
+import GHC.Generics (Generic)
+
+-- | Signed values are either positive or negative.
+data Signed a = Positive a | Negative a
+  deriving (Eq, Ord, Generic, Show, Read)
+
+instance Functor Signed where
+  fmap f (Positive x) = Positive (f x)
+  fmap f (Negative x) = Negative (f x)
+
+instance Traversable Signed where
+  traverse f (Positive x) = Positive <$> f x
+  traverse f (Negative x) = Negative <$> f x
+
+instance Foldable Signed where
+  foldMap = foldMapDefault
+
+instance Applicative Signed where
+  pure = Positive
+  (<*>) = ap
+
+instance Monad Signed where
+  Positive x >>= f = f x
+  Negative x >>= f = negateSigned $ f x
+
+infix 9 /\
+infix 9 \/
+
+-- | A boolean type class.
+class Boolean f where
+  (/\) :: f a -> f a -> f a
+  (\/) :: f a -> f a -> f a
+  bNot :: f a -> f a
+  bTrue :: f a
+  bFalse :: f a
+  bConst :: Signed a -> f a
+
+-- | Syntax of boolean expressions parameterized over a
+-- set of leaves, named constants.
+data BoolExpr a
+  = BAnd (BoolExpr a) (BoolExpr a)
+  | BOr (BoolExpr a) (BoolExpr a)
+  | BNot (BoolExpr a)
+  | BTrue
+  | BFalse
+  | BConst (Signed a)
+  deriving (Eq, Ord, Generic, Show {-! derive : Arbitrary !-})
```

Comment:
I am confused - where are `ToJSON` and `FromJSON` of `BoolExpr`? 

They should be easy to craft by hand and write here.

Reply:
The JSON is parsed as the `Prerequsite` type. I'm only using the `BoolExpr` type internally, after converting from the `Prerequsite` the type with `toBoolExpr`.